[SourceKit] Include generated macro buffers in diagnostic responses

hamishknight opened the pull request:
Introduce a new key `generated_buffers`, which stores an array of generated buffers. These include the buffer text, as well as its original location and any parent buffers.

While here, also fix rdar://107281079 such that only apply the filename fallback logic to the pretty-printed Decl case. We ought to remove this fallback once the editor can handle it though.

rdar://107281079
rdar://107952288

hamishknight commented:
Still need to add a test case for rdar://107281079

hamishknight commented:
@swift-ci please test

hamishknight commented on test/SourceKit/DocumentStructure/structure.swift.empty.response:
```diff
@@ -1671,14 +1671,20 @@
     }
   ],
+  key.buffer_infos: [
+    {
+      key.buffer_name: ""
+    }
```
Should we be writing out an empty buffer name here? Or should we just drop the key like we do with `filename`? Seems weird to just have an empty object though.

hamishknight commented on test/SourceKit/DocumentStructure/structure.swift.empty.response:
```diff
@@ -1692,7 +1698,8 @@
           key.length: 30
         }
-      ]
+      ],
+      key.buffer_at_index: 0
```
Note currently we're generating buffer info regardless of whether the buffer is for a macro expansion. IMO it would be nice if we could deprecate `filename` and just have clients refer to the buffer info, which would save needing to write the filename N times for N diagnostics in the same file.

hamishknight commented on tools/SourceKit/lib/SwiftLang/SwiftEditor.cpp:
```diff
@@ -64,15 +64,74 @@ void EditorDiagConsumer::getAllDiagnostics(
+          auto OrigRawRange = getRawRangeInBuffer(OrigRange, OrigBuffer, SM);
+          assert(OrigRawRange);
+
+          auto OrigFileName = SM.getDisplayNameForLoc(OrigLoc);
+          auto OrigInfo = getBufferInfo(OrigFileName, OrigBuffer, SM);
+          OriginalLocInfo.emplace(OrigInfo, OrigBuffer, *OrigRawRange);
```
This recursion assumes original locations forms a DAG, is that a safe assumption to make? We could eagerly create and insert the buffer info if we want to be defensive here.

bnbarham replied:
I can't imagine a case that it wouldn't be, seems fine to me.

bnbarham commented on tools/SourceKit/lib/SwiftLang/SwiftEditor.cpp:
```diff
@@ -64,15 +64,74 @@ void EditorDiagConsumer::getAllDiagnostics(
+    if (auto Info = SM.getGeneratedSourceInfo(*BufferID)) {
+      // Don't include this information for replaced function body buffers, as
+      // they'll just be referencing the original file.
+      // FIXME: Does this mean that the location information we'll be giving
```
`ReplacedFunctionBody` only happens for completion/cursor info at the moment. So we should never end up with that case here. You could add an assertion if you want.

hamishknight replied:
We actually hit this case when doing multiple `compile` requests

bnbarham replied:
Oh interesting. @rintaro despite you just talking `compile` I didn't think about it here. It does seem like we would be messing up diagnostics then, I'm unclear whether we actually care though - I don't believe we send them back.

rintaro replied:
Yes. But I didn't think much about diagnostics for `compile` requests. In our intended use cases, I thought we only wanted to know if there's any error or not (succeeded or not). But anyway `compile` request hasn't been maintained well, and probably requires major refinements.

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test

ahoppen reviewed (approved):
Looks reasonable to me

hamishknight commented on tools/SourceKit/lib/SwiftLang/SwiftEditor.cpp:
```diff
@@ -64,15 +64,75 @@ void EditorDiagConsumer::getAllDiagnostics(
+      // FIXME: Does this mean that the location information we'll be giving
+      // will be wrong? Seems like we may need to adjust it to correspond to the
+      // original file.
+      if (Info->kind != GeneratedSourceInfo::ReplacedFunctionBody) {
```
I believe it's used for cached SourceKit operations where we re-use an existing AST, but splice in the updated text for a function body. The new text is stored in this kind of buffer.

bnbarham replied:
Yeah, it's just the function body replacing for completion (and now cursor info)... and the compile case that I forgot about :)

ahoppen replied:
Ah, makes sense.

hamishknight commented on test/SourceKit/Inputs/sourcekitd_path_sanitize.py:
```diff
@@ -12,21 +12,35 @@
-HEADER_RE = re.compile(r' file=\\".*[/\\](.*)\.h\\"')
+# I'm sorry dear reader, unfortunately my knowledge of regex trumps my knowledge of
+# Python. This pattern allows us to clean up file paths by stripping them down to
+# just the relevant file name.
```
I'm cursed with knowledge 😄

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test Windows

hamishknight commented:
@swift-ci please test Windows

bnbarham commented on test/SourceKit/Diagnostics/rdar107281079.swift:
```diff
@@ -0,0 +1,32 @@
+// RUN: %sourcekitd-test -req=diags %s -- %s -I %S/../Inputs/header_with_macro | %FileCheck %s
```
Test name is fairly opaque to OSS 😅. `cross_language_diagnostic.swift` or something?

bnbarham commented on test/SourceKit/Inputs/header_with_macro/Header.h:
```diff
@@ -0,0 +1,6 @@
+#ifndef Header_h
```
You can keep it in Inputs if you want, but I much prefer using `split-file`. I admit it's a little weird when cross language though.

hamishknight replied:
Hmm yeah, it being cross language was the main reason I didn't use `split-file`, didn't want to confuse the editor by putting C code in a swift file. That being said, I guess this is a simple enough case where it doesn't matter too much.

hamishknight commented:
@swift-ci please test

hamishknight committed:
Avoid computing generated source notes on child notes

This could cause us to double up on the same note
for nested macro expansions, as we'd generate the
note, then when recursing back through the function
we'd compute the same note again. Also remove the
seemingly unused `lastBufferID` param.

hamishknight committed:
[SourceKit] Include generated macro buffers in diagnostic responses

Introduce a new key `generated_buffers`, which
stores an array of generated buffers. These
include the buffer text, as well as its original
location and any parent buffers.

While here, also fix rdar://107281079 such that
only apply the filename fallback logic to the
pretty-printed Decl case. We ought to remove this
fallback once the editor can handle it though.

rdar://107281079
rdar://107952288

hamishknight committed:
[SourceKit] Introduce RawCharSourceRange

Use this to simplify the writing of ranges into
SourceKit responses.

hamishknight committed:
NFC: Move some code around

hamishknight commented:
@swift-ci please test

hamishknight commented:
Yeah, I [updated the tests](https://github.com/apple/swift/compare/871441541a180d348f8618143ec93c74187b3830..968ceb3d748fe5ec324cbcb6ddfe599e546ac7dc)
